### PR TITLE
Bug 1267247 - Add U2F v1.1 support to Firefox for iOS

### PR DESCRIPTION
**Pull request by bifurcation**

### CODE REVIEW

File: Client/Assets/U2F.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    });
+
+    Object.defineProperty(window, "u2f", {
+        writable: false,
+        value: u2f
+    });
+
+    window.__firefox__.u2f = {
+    finish: function finishRegister(obj) {
```
Comment (jcjones) [inline Client/Assets/U2F.js]: nit: This seems like it should be called something else as it also clean up after "sign".

File: Client/U2FHelper.swift [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+    }
+
+    private func authenticate(result: (Bool -> ())) {
+        // We manage the TouchID prompt ourselves (instead of using, say AppAuthenticator)
+        // because we really want it to appear on every call.
+        let context = LAContext()
+        let policy = LAPolicy.DeviceOwnerAuthenticationWithBiometrics
+        if (context.canEvaluatePolicy(policy, error: nil)) {
+            context.evaluatePolicy(policy, localizedReason: kAuthenticationReason, reply:{ authenticated, error in
+                if error != nil {
```
Comment (jcjones) [inline Client/U2FHelper.swift]: It'd probably be valuable for debugging to be able to log that error someplace, but I don't know how the logging works in firefox-ios.

File: Client/U2FHelper.swift [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+        var responseDataPrefix : UInt8 = 0x05
+        var keyHandleLen : UInt8 = UInt8(keyHandle.length)
+        responseData.appendBytes(&responseDataPrefix, length: 1)
+        responseData.appendData(pubBytes)
+        responseData.appendBytes(&keyHandleLen, length: 1)
+        responseData.appendData(keyHandle)
+        responseData.appendData(attestationCert)
+        responseData.appendData(attestationSig)
+
+        result(responseData.base64URLEncodedStringWithOptions(NSDataBase64EncodingOptions()))
```
Comment (jcjones) [inline Client/U2FHelper.swift]: Should we zeroize `privBytes` when it falls out of scope here?

File: Client/U2FHelper.swift [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+        message.appendData(appParam)
+        message.appendData(presenceAndCounter)
+        message.appendData(challengeParam)
+        let sig = privateKey.signMessage(message)
+
+        // Assemble the response data
+        let responseData = presenceAndCounter
+        responseData.appendData(sig)
+        result(responseData.base64URLEncodedStringWithOptions(NSDataBase64EncodingOptions()))
+    }
```
Comment (jcjones) [inline Client/U2FHelper.swift]: Similar to the above, should we zeroize `privB64` and `privBytes` on our way out?

File: FxA/FxA/ECDSAKeyPair.m [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+            if (ctx != NULL) {
+                if (EVP_VerifyInit_ex(ctx, EVP_sha256(), NULL) == 1) {
+                    if (EVP_VerifyUpdate(ctx, [message bytes], [message length]) ==  1) {
+                        int err = EVP_VerifyFinal(ctx, [signature bytes], [signature length], pkey);
+                        if (err == 1) {
+                            verified = YES;
+                        } else if (err == -1) {
+                            unsigned long e = ERR_get_error();
+                            
+                            char buf[120];
```
Comment (jcjones) [inline FxA/FxA/ECDSAKeyPair.m]: [OpenSSL Documentation](https://www.openssl.org/docs/manmaster/crypto/ERR_error_string.html) defines this `buf` as being at least 256 bytes long.

If you want to use the shorter buffer, please use `ERR_error_string_n`

File: SyncTests/CryptoTests.swift [hunk 6]
```diff
@@ -80,4 +81,63 @@ class CryptoTests: XCTestCase {
...
+        let privECDSA2 = ECDSAPrivateKey(binaryRepresentation: privBytes, group: .P256)
+        XCTAssertNotNil(privECDSA2)
+        let sigECDSA2 = privECDSA2.signMessage(cleartext)
+        let verECDSA2 = ecdsa.publicKey.verifySignature(sigECDSA2, againstMessage: cleartext)
+        XCTAssertTrue(verECDSA2)
+
+        let pubBytes = ecdsa.publicKey.BinaryRepresentation()
+        let pubECDSA3 = ECDSAPublicKey(binaryRepresentation: pubBytes, group: .P256)
+        XCTAssertNotNil(pubECDSA3)
+        let verECDSA3 = pubECDSA3.verifySignature(sigECDSA, againstMessage: cleartext)
```
Comment (jcjones) [inline SyncTests/CryptoTests.swift]: Nit: It's sort of weird that this `verifySignature` doesn't check `sigECDSA2`, but it's NBD.

File: SyncTests/CryptoTests.swift [hunk 7]
```diff
@@ -80,4 +81,63 @@ class CryptoTests: XCTestCase {
...
+        XCTAssertTrue(verECDSA3)
+
+        // Just test that this runs and produces non-empty output
+        let cert = ecdsa.privateKey.selfSignedCertificateWithName("Test cert", slack: 0, lifetime: 60 * 60)
+        XCTAssertNotNil(cert)
```
Comment (jcjones) [inline SyncTests/CryptoTests.swift]: Nit: If there is any cert parsing code handy, it'd be good to assert parsability, but that could be a follow-up.

**Comment by jcjones** [comment]
I've completed a crypto / spec review of this code; several nits, only a couple non-nitty things. I should probably take another look after any rework from the iOS review though.

**Comment by farhanpatel** [comment]
Thanks for this PR! We just finished a major migration to swift 3.0 making merging in most patches pretty difficult. If you feel like this patch is still important please feel free to update your PR to swift 3 and reopen this PR 😄

**Pull request closed without merging**